### DISCUSSION TITLE
Confused about "focus"?

### DISCUSSION THREAD

**Discussion by voidware** [ocornut/imgui#3945] [Old/Archived discussions]
I'd like to ask some novice questions;

What does it mean for an item to have the "focus"? Is this the same as being "active"? Is "focus" related to keyboard focus or is it just a question of what is highlighted.

the demo has code like this:

```
        // Command-line
        bool reclaim_focus = false;
        ImGuiInputTextFlags input_text_flags = ...;
        if (ImGui::InputText("Input", InputBuf, IM_ARRAYSIZE(InputBuf), input_text_flags, &TextEditCallbackStub, (void*)this))
        {
            /// ...
            reclaim_focus = true;
        }

        // Auto-focus on window apparition
        ImGui::SetItemDefaultFocus();
        if (reclaim_focus)
            ImGui::SetKeyboardFocusHere(-1); // Auto focus previous widget
```

What does `SetItemDefaultFocus();` do here?

Thanks for helping.

**Comment by ocornut** [comment]
There are legacy issues involved here, hence the confusion:

- "Focus" should be Navigation (Keyboard/Gamepad) focus aka current highlighted widget when navigation is active.
- "Active" is when a button is being held or a text field is being edited.

- The `SetKeyboardFocusHere()` function does **BOTH**, AND, it only works on widgets that are declaring themselves of the old Tabbing system. So essentially `SetKeyboardFocusHere()` is at fault here for being both confusing (name wise) and incomplete feature wise.

Among some of one million unfinished stashes, I have once started working on tidying up all that and new API, which would more clearly distinguish Focus and Activation, e.g. `FocusItem()` vs `ActivateItem()`.
`SetKeyboardFocusHere()` would then be a legacy function for e.g. `Activate next item`.

**Reply by voidware** [reply]
Aha! Thanks for this.

So `SetItemDefaultFocus` isnt needed except to help the tabbing system. and "here" means the next item not the one just created.

Got it. thanks.
